### PR TITLE
expression: rewrite builtin function: ISNULL

### PR DESCRIPTION
**Pull request by breeswish**
to https://github.com/pingcap/tidb/issues/4080

@XuHuaiyu @zz-jason

### CODE REVIEW

File: expression/builtin_op.go [hunk 1]
```diff
@@ -659,25 +664,113 @@ type isNullFunctionClass struct {
...
+		sig = &builtinIntIsNullSig{baseIntBuiltinFunc{bf}}
+	case tpDecimal:
+		sig = &builtinDecimalIsNullSig{baseIntBuiltinFunc{bf}}
+	case tpReal:
+		sig = &builtinRealIsNullSig{baseIntBuiltinFunc{bf}}
+	case tpTime:
+		sig = &builtinTimeIsNullSig{baseIntBuiltinFunc{bf}}
+	case tpDuration:
+		sig = &builtinDurationIsNullSig{baseIntBuiltinFunc{bf}}
+	case tpString:
```
Comment (zz-jason) [inline expression/builtin_op.go]: since `default` is never reached, you can `s/case tpString:/default:/`

**Comment by hanfei1991** [comment]
LGTM

File: expression/builtin_op.go [hunk 2]
```diff
@@ -659,25 +664,113 @@ type isNullFunctionClass struct {
 }
 
 func (c *isNullFunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
-	sig := &builtinIsNullSig{newBaseBuiltinFunc(args, ctx)}
-	return sig.setSelf(sig), errors.Trace(c.verifyArgs(args))
+	if err := c.verifyArgs(args); err != nil {
+		return nil, errors.Trace(err)
+	}
+	tc := args[0].GetType().ToClass()
+	var argTp evalTp
```
Comment (XuHuaiyu) [inline expression/builtin_op.go]: extract a function like `fieldTp2EvalTp`.

Comment (breeswish) [inline expression/builtin_op.go]: Let's do it in another PR since the rest of code has similar patterns.

File: expression/builtin_op.go (hunk 1 above)
Comment (breeswish) [inline expression/builtin_op.go]: I would like to keep this for clearer purpose. In addition, if we add another evalTp in the future, it will fall into the `default` cluster, which may not be what we want.

File: expression/builtin_op.go (hunk 2 above)
Comment (XuHuaiyu) [inline expression/builtin_op.go]: ok
